pkg/apis/ceph.rook.io/v1beta1/types.go

```diff
@@ -170,13 +170,16 @@ type FilesystemList struct {
 // FilesystemSpec represents the spec of a file system
 type FilesystemSpec struct {
 	// The metadata pool settings
-	MetadataPool PoolSpec `json:"metadataPool"`
+	MetadataPool PoolSpec `json:"metadataPool,omitempty"`
 
 	// The data pool settings
-	DataPools []PoolSpec `json:"dataPools"`
+	DataPools []PoolSpec `json:"dataPools,omitempty"`
 
 	// The mds pod info
 	MetadataServer MetadataServerSpec `json:"metadataServer"`
+
+	// Don't touch pools/filesystem, just run MDS daemons
+	OnlyManageDaemons bool `json:"onlyManageDaemons"`
```

Comment:
I assume if this flag is `true` the values for `DataPools` and `MetadataPools` must be empty, correct? I wonder if it would be cleaner to define another `FileSystemSpec` like so:

```
type FilesystemPoolSpec struct {
 	// The metadata pool settings
 	MetadataPool PoolSpec `json:"metadataPool"`
 	MetadataPool PoolSpec `json:"metadataPool,omitempty"`

 	// The data pool settings
 	DataPools []PoolSpec `json:"dataPools"`
 	DataPools []PoolSpec `json:"dataPools,omitempty"`
}

// FilesystemSpec represents the spec of a file system
type FilesystemSpec struct {
	// The data and metadata pool settings. If empty Rook will not create or manage pools
	FilesystemPoolSpec FilesystemPoolSpec `json:"pools,omitempty"`

	// The mds pod info
	MetadataServer MetadataServerSpec `json:"metadataServer"`
}
```

Reply:
Why would we require the pool specs to be empty? If that setting is true, it just means the operator would ignore them if they are set. We have several other places in the CRDs where settings are ignored depending on other settings. For example, if `useAllNodes: true`, the entire `nodes` element is ignored.

Reply:
The pools do not have to be empty.  In practice, they would probably be non-empty if a user initially created their pools using Rook, and then set onlyManageDaemons=true right before deleting their Rook filesystem, to preserve their data.  They *could* also clear the pools element, but I don't see any reason to force them to.

Reply:
what does it mean for `OnlyManageDaemons` to be set true and the pool specs to be specified? I'm attempting to find a clean way to delineate between a declaratively managed an unmanaged resource.

>  if `useAllNodes: true`, the entire nodes element is ignored.

that should be a validation error, right? if `useAllNodes` is true, `nodes` should be omitted. you can specify these kinds of validation rules now in CRD specs.

>  if a user initially created their pools using Rook, and then set onlyManageDaemons=true right before deleting their Rook filesystem, to preserve their data.

this is the kind of partial management that concerns me. I.e. create the pools with rook and then rook stops managing them due to a flag

Reply:
> you can specify these kinds of validation rules now in CRD specs.

Although it feels a bit overkill, I don't have any fundamental problem making this stricter -- could you help me out with a pointer to how to do this stuff with k8s CRDs?

Reply:
https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation

Reply:
We still need overall CRD validation with #1000. What if we add this check at the same time that broader issue is resolved?

Reply:
postponing validation is OK with me. my preference above is to go with a spec that make it clearer that rook is not managing pools at all (see spec above). does that work?

Reply:
Is the same goal not accomplished with the `omitempty` flag on the `metadataPool` and `dataPools`? I do see that your suggestion could be clearer, but CRD migration has a non-trivial cost.

Reply:
you can achieve the same goal with `omitempty` and eventually adding CRD validation to make sure both are specified or both are omitted.

We would not need `OnlyManageDaemons` in this approach and this takes us back to the original suggestion earlier in the PR :-)

Reply:
Agreed, we could certainly use the resource boundary in this case. I was preferring the explicit setting because it has implications not just for skipping pool management, but also for skipping creating the filesystem resource. Currently the filesystem resource has no settings so is not visible in the crd to the user, so it seemed awkward to skip creating it based on not specifying pools. But since it's hidden from the user anyway, it seems reasonable to infer.